File: src/main/java/com/ultreon/bubbles/entity/ammo/BasicAmmoType.java
package com.ultreon.bubbles.entity.ammo;

import com.ultreon.bubbles.common.AttributeMap;
import com.ultreon.bubbles.entity.AmmoEntity;
import com.ultreon.bubbles.entity.attribute.Attribute;

import java.awt.*;
import com.ultreon.hydro.render.Renderer;
import java.awt.geom.AffineTransform;
import java.awt.geom.Area;
import java.awt.geom.Line2D;

public class BasicAmmoType extends AmmoType {
    private final Line2D line = new Line2D.Double(0, 0, 5, 0);

    @Override
    public Shape getShape(AmmoEntity entity) {
        Area area = new Area(line);

        AffineTransform transform = new AffineTransform();
        transform.rotate(Math.toRadians(entity.getRotation()), area.getBounds().getCenterX(), area.getBounds().getCenterY());

        area.transform(transform);

        return line;
    }

    @Override
    public void render(Renderer g, AmmoEntity entity) {
        Renderer g2 = (Renderer) g;
        g2.fill(getShape(entity));

    }

    @Override
    public AttributeMap getDefaultAttributes() {
        AttributeMap map = new AttributeMap();
        map.setBase(Attribute.ATTACK, 1f);
        map.setBase(Attribute.DEFENSE, 4f);
        return map;
    }
}

File: src/main/kotlin/qtech/bubbles/common/ammo/BasicAmmoType.kt
package qtech.bubbles.common.ammo

import qtech.bubbles.common.AttributeMap
import qtech.bubbles.common.entity.Attribute
import qtech.bubbles.entity.AmmoEntity
import qtech.hydro.GraphicsProcessor
import java.awt.Shape
import java.awt.geom.AffineTransform
import java.awt.geom.Area
import java.awt.geom.Line2D

class BasicAmmoType : AmmoType() {
    private val line: Line2D = Line2D.Double(0.0, 0.0, 5.0, 0.0)
    override fun getShape(ammoEntity: AmmoEntity): Shape {
        val area = Area(line)
        val transform = AffineTransform()
        transform.rotate(Math.toRadians(ammoEntity.rotation), area.bounds.centerX, area.bounds.centerY)
        area.transform(transform)
        return line
    }

    override fun render(g: GraphicsProcessor, ammoEntity: AmmoEntity) {
        g.shapeF(getShape(ammoEntity))
    }

    override val defaultAttributes: AttributeMap
        get() {
            val map = AttributeMap()
            map.set(Attribute.ATTACK, 1f)
            map.set(Attribute.DEFENSE, 4f)
            return map
        }
}